File: app/src/main/java/knf/kuma/tv/search/TVSearchFragment.java
package knf.kuma.tv.search;

import android.Manifest;
import android.content.pm.PackageManager;
import android.os.Bundle;

import java.util.List;

import androidx.core.content.ContextCompat;
import androidx.leanback.app.SearchSupportFragment;
import androidx.leanback.widget.ArrayObjectAdapter;
import androidx.leanback.widget.HeaderItem;
import androidx.leanback.widget.ListRow;
import androidx.leanback.widget.ListRowPresenter;
import androidx.leanback.widget.ObjectAdapter;
import androidx.leanback.widget.OnItemViewClickedListener;
import androidx.leanback.widget.Presenter;
import androidx.leanback.widget.Row;
import androidx.leanback.widget.RowPresenter;
import androidx.leanback.widget.SpeechRecognitionCallback;
import androidx.lifecycle.LiveData;
import knf.kuma.database.CacheDB;
import knf.kuma.pojos.AnimeObject;
import knf.kuma.tv.anime.AnimePresenter;
import knf.kuma.tv.details.TVAnimesDetails;

public class TVSearchFragment extends SearchSupportFragment implements SearchSupportFragment.SearchResultProvider, SpeechRecognitionCallback, OnItemViewClickedListener {
    private ArrayObjectAdapter arrayObjectAdapter;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        checkPermissions();
        arrayObjectAdapter = new ArrayObjectAdapter(new ListRowPresenter());
        setSearchResultProvider(this);
        setOnItemViewClickedListener(this);
        if (getContext().getPackageManager().hasSystemFeature("amazon.hardware.fire_tv"))
            setSpeechRecognitionCallback(this);
    }

    private void checkPermissions() {
        if (ContextCompat.checkSelfPermission(getContext(), Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED)
            requestPermissions(new String[]{Manifest.permission.RECORD_AUDIO}, 55498);
    }

    @Override
    public ObjectAdapter getResultsAdapter() {
        return arrayObjectAdapter;
    }

    @Override
    public boolean onQueryTextChange(String newQuery) {
        setResult(newQuery);
        return true;
    }

    @Override
    public void recognizeSpeech() {

    }

    @Override
    public boolean onQueryTextSubmit(String query) {
        setResult(query);
        return true;
    }

    private void setResult(String query) {
        final LiveData<List<AnimeObject>> liveData = CacheDB.INSTANCE.animeDAO().getSearchList("%" + query + "%");
        if (getActivity() != null) {
            liveData.observe(getActivity(), animeObjects -> {
                liveData.removeObservers(getActivity());
                arrayObjectAdapter.clear();
                ArrayObjectAdapter objectAdapter = new ArrayObjectAdapter(new AnimePresenter());
                for (AnimeObject object : animeObjects)
                    objectAdapter.add(object);
                HeaderItem headerItem = new HeaderItem(animeObjects.size() > 0 ? "Resultados para '" + query + "'" : "Sin resultados");
                arrayObjectAdapter.add(new ListRow(headerItem, objectAdapter));
            });
        }
    }

    @Override
    public void onItemClicked(Presenter.ViewHolder itemViewHolder, Object item, RowPresenter.ViewHolder rowViewHolder, Row row) {
        AnimeObject object = (AnimeObject) item;
        TVAnimesDetails.start(getContext(), object.link);
    }
}

File: app/src/main/java/knf/kuma/tv/search/TVSearchFragment.kt
package knf.kuma.tv.search

import android.os.Bundle
import android.view.View
import androidx.leanback.app.SearchSupportFragment
import androidx.leanback.paging.PagingDataAdapter
import androidx.leanback.widget.ArrayObjectAdapter
import androidx.leanback.widget.HeaderItem
import androidx.leanback.widget.ListRow
import androidx.leanback.widget.ListRowPresenter
import androidx.leanback.widget.ObjectAdapter
import androidx.leanback.widget.OnItemViewClickedListener
import androidx.leanback.widget.Presenter
import androidx.leanback.widget.Row
import androidx.leanback.widget.RowPresenter
import androidx.lifecycle.LiveData
import androidx.lifecycle.Observer
import androidx.lifecycle.lifecycleScope
import androidx.paging.Pager
import androidx.paging.PagingConfig
import com.inmobi.media.la
import knf.kuma.database.CacheDB
import knf.kuma.pojos.av1.DirectoryAV1
import knf.kuma.pojos.av1.DirectoryAV1Min
import knf.kuma.pojos.av1.Genre
import knf.kuma.pojos.av1.SearchDataSource
import knf.kuma.search.SearchFragmentMaterial
import knf.kuma.tv.TVRepository
import knf.kuma.tv.anime.AnimePresenter
import knf.kuma.tv.details.TVAnimesDetails
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.mapLatest
import kotlinx.coroutines.launch

class TVSearchFragment : SearchSupportFragment(), SearchSupportFragment.SearchResultProvider, OnItemViewClickedListener {
    private var arrayObjectAdapter: ArrayObjectAdapter? = null
    private val adapter by lazy { PagingDataAdapter(AnimePresenter(), DirectoryAV1Min.DIFF) }
    private var resultsRow: Row? = null

    private val currentQuery = MutableStateFlow("")

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        arrayObjectAdapter = ArrayObjectAdapter(ListRowPresenter())
        setSearchResultProvider(this)
        setOnItemViewClickedListener(this)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        arrayObjectAdapter?.clear()
        arrayObjectAdapter?.add(ListRow(HeaderItem("Géneros"), ArrayObjectAdapter(TagPresenter()).also {
            it.addAll(0, SearchFragmentMaterial.genres)
        }))
        resultsRow = ListRow(HeaderItem("Todos los animes"), adapter).also {
            arrayObjectAdapter?.add(it)
        }
        viewLifecycleOwner.lifecycleScope.launch {
            currentQuery.flatMapLatest {
                TVRepository.searchQuery(it)
            }.collectLatest {
                adapter.submitData(it)
                val headerItem = HeaderItem(
                    when {
                        currentQuery.value.isEmpty() -> "Todos los animes"
                        else -> "Resultados para '${currentQuery.value}'"
                    }
                )
                resultsRow?.headerItem = headerItem
            }
        }
    }

    override fun getResultsAdapter(): ObjectAdapter? {
        return arrayObjectAdapter
    }

    override fun onQueryTextChange(newQuery: String): Boolean {
        currentQuery.value = newQuery
        return true
    }

    override fun onQueryTextSubmit(query: String): Boolean {
        currentQuery.value = query
        return true
    }

    override fun onItemClicked(itemViewHolder: Presenter.ViewHolder, item: Any, rowViewHolder: RowPresenter.ViewHolder, row: Row) {
        when (item) {
            is DirectoryAV1Min -> context?.let { TVAnimesDetails.start(it, item.animeUrl) }
            is Genre -> context?.let { TVTag.start(it, item) }
        }
    }
}
